src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.Collector.cs

```diff
@@ -0,0 +1,463 @@
+            public void HandleDeclarationExpression(IDeclarationExpressionOperation op)
+            {
+                if (op.Expression != null && CanUpgrade(op.Expression))
+                {
+                    if (op.Expression.Kind == OperationKind.LocalReference)
+                    {
+                        var localRef = (ILocalReferenceOperation)op.Expression;
+                        if (localRef.Type != null)
+                        {
+                            var bag = LocalAssignments.GetOrAdd(localRef.Local, _ => new ConcurrentBag<ITypeSymbol>());
+                            bag.Add(localRef.Type);
+                        }
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Record the type of values returned by a given method.
+            /// </summary>
+            public void HandleReturn(IReturnOperation op)
+            {
+                if (op.ReturnedValue != null)
+                {
+                    if (op.SemanticModel!.GetEnclosingSymbol(op.Syntax.SpanStart) is IMethodSymbol methodSym && CanUpgrade(methodSym))
+                    {
+                        var valueTypes = GetValueTypes(op.ReturnedValue);
+                        foreach (var valueType in valueTypes)
+                        {
+                            RecordAssignment(methodSym, valueType);
+                        }
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Trivial reject for types that can't be upgraded in order to avoid wasted work.
+            /// </summary>
+            private static bool CanUpgrade(IOperation target)
+                => target.Type == null || (!target.Type.IsSealed && !target.Type.IsValueType);
+
+            /// <summary>
+            /// Trivial reject for methods that can't be upgraded in order to avoid wasted work.
+            /// </summary>
+            private static bool CanUpgrade(IMethodSymbol methodSym)
+                => methodSym.DeclaredAccessibility == Accessibility.Private && methodSym.MethodKind == MethodKind.Ordinary;
+
+            /// <summary>
+            /// Trivial reject for fields that can't be upgraded in order to avoid wasted work.
+            /// </summary>
+            private static bool CanUpgrade(IFieldSymbol fieldSym)
+                => fieldSym.DeclaredAccessibility == Accessibility.Private;
+
+            private List<ITypeSymbol> GetValueTypes(IOperation op)
+            {
+                var values = new List<ITypeSymbol>();
+                GetValueTypes(values, op);
+                return values;
+            }
+
+            private void GetValueTypes(List<ITypeSymbol> values, IOperation op)
+            {
+                switch (op.Kind)
+                {
+                    case OperationKind.Literal:
+                        {
+                            var lit = (ILiteralOperation)op;
+                            if (lit.Type == null)
```

Comment:
You should instead use this extension method: https://github.com/dotnet/roslyn-analyzers/blob/460a0e5d36934353e88a5e04cdb39bcc0d0b6545/src/Utilities/Compiler/Extensions/IOperationExtensions.cs#L61-L64

Reply:
Done.